src/main/java/uk/gov/hmcts/reform/blobrouter/services/EnvelopeActionService.java

```diff
@@ -0,0 +1,87 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import uk.gov.hmcts.reform.blobrouter.data.envelopes.Envelope;
+import uk.gov.hmcts.reform.blobrouter.data.envelopes.EnvelopeRepository;
+import uk.gov.hmcts.reform.blobrouter.data.envelopes.Status;
+import uk.gov.hmcts.reform.blobrouter.data.events.EnvelopeEventRepository;
+import uk.gov.hmcts.reform.blobrouter.data.events.NewEnvelopeEvent;
+import uk.gov.hmcts.reform.blobrouter.exceptions.EnvelopeCompletedOrNotStaleException;
+import uk.gov.hmcts.reform.blobrouter.exceptions.EnvelopeNotFoundException;
+
+import java.time.Instant;
+import java.util.UUID;
+
+import static java.time.Duration.between;
+import static java.time.Instant.now;
+import static java.util.Comparator.naturalOrder;
+import static uk.gov.hmcts.reform.blobrouter.data.envelopes.Status.DISPATCHED;
+import static uk.gov.hmcts.reform.blobrouter.data.envelopes.Status.REJECTED;
+import static uk.gov.hmcts.reform.blobrouter.data.events.ErrorCode.ERR_STALE_ENVELOPE;
+import static uk.gov.hmcts.reform.blobrouter.data.events.EventType.MANUALLY_REJECTED;
+
+@Service
+public class EnvelopeActionService {
+
+    private final EnvelopeRepository envelopeRepository;
+    private final EnvelopeEventRepository envelopeEventRepository;
+    private final long creationTimeoutHr;
+
+    public EnvelopeActionService(
+        EnvelopeRepository envelopeRepository,
+        EnvelopeEventRepository envelopeEventRepository,
+        @Value("${creation-stale-timeout-hr}") long creationTimeoutHr
+    ) {
+        this.envelopeRepository = envelopeRepository;
+        this.envelopeEventRepository = envelopeEventRepository;
+        this.creationTimeoutHr = creationTimeoutHr;
+    }
+
+    @Transactional
+    public void rejectStaleEnvelope(UUID envelopeId) {
+        Envelope envelope = envelopeRepository.find(envelopeId)
+            .orElseThrow(
+                () -> new EnvelopeNotFoundException("Envelope with id " + envelopeId + " not found")
+            );
+        validateEnvelopeState(envelope);
+
+        createEvent(envelope);
+
+        envelopeRepository.updateStatus(envelopeId, REJECTED);
+    }
+
+    private void createEvent(Envelope envelope) {
+        NewEnvelopeEvent event = new NewEnvelopeEvent(
+            envelope.id,
+            MANUALLY_REJECTED,
+            ERR_STALE_ENVELOPE,
+            "Manually rejected due to stale state"
+        );
+        envelopeEventRepository.insert(event);
+    }
+
+    private void validateEnvelopeState(Envelope envelope) {
+        if (envelope.status == DISPATCHED || envelope.status == REJECTED
+            || !isStale(envelope)) {
+            throw new EnvelopeCompletedOrNotStaleException(
+                "Envelope with id " + envelope.id + " is completed or not stale"
+            );
+        }
+    }
+
+    private boolean isStale(Envelope envelope) {
+        if (envelope.status != Status.CREATED) {
```

Comment:
addition to this, if it is CFT envelope can we check processor?

Reply:
Processor and router have different databases. As mentioned in https://tools.hmcts.net/jira/browse/DTSBPS-348?focusedCommentId=952256&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-952256 there are several zip files in router which are not known by processor. This means that we need to check and handle in processor separately.

Reply:
there is an end point to check

Reply:
The idea here is to handle an envelope which is stale in router i.e. in CREATED status with after e.g. 2 hours. It does not matter if it exists in processor.

Reply:
if file is processed in processor successfully you can not mark as rejected?